### DISCUSSION TITLE
It would be nice to have useAtomValue lint for async atom

### DISCUSSION THREAD

**Discussion by devnomic** [pmndrs/jotai#1899] [General] [answered]
Just  an idea / suggestion, if it's possible to have lint when using `useAtomValue` with async atom. 
If the component is not wrapped in suspense lint will warn.

**Reason** 
When refactoring to update sync atom to async atom, we could miss `useAtomValue` that not wrapped in suspense especially if the codebase is quite big.

**Comment by dai-shi** [comment]
I don't think it's possible for lint to warn such things.


---

fwiw, there's an undocumented feature to avoid suspending for immediately resolving promises:
https://github.com/pmndrs/jotai/blob/94c1a20bc38ece90b75155571c71eb43e07303cd/src/react/useAtomValue.ts#L45

**Comment by devnomic** [comment]
Wow, thank you! another thing i didn't know exist. So basically `delay` suitable for managing short asynchronous operations that we don't want to write suspense for it?

Back to linting,

**Case**
I need to refactor codebase,
The problem is i have some variables that read from `localStorage` which is sync atom.
Now i need to read from another place which is async.

Since in jotai we define from smallest atom first and derive or use it to make more complex atoms. this will trigger chains of atoms i need to update.

In this case, the smallest atom is variables that reading from `localstorage` need to be async now.
I can refactor all derived atoms with confidence since the IDE will tell me which atoms need to be updated (previously sync, now you need to write`await`).

But this was not the case with the Component. There is no IDE complain on `useAtomValue` since it can use sync and async atom.
There will be crash on runtime only. Ideally I need to know beforehand (kind of compile time error) when refactor.
It really painful to check one by one, could miss something.

I need the IDE to tell me, look you are using `useAtomValue` with async value, you need to wrap this with suspense.
But since this is not possible with eslint, I'm thinking so the only way to have information on which variable use sync or async atom is to separate them, using another util functions such as:

`useSyncAtomValue` (can only get from sync atom)
`useAsyncAtomValue` (can only get from async atom)

This will make IDE complain, hey you are using `useSyncAtomValue` but your atom is `async`.
Then i can refactor with confidence since async atom always trigger suspense, i will make sure the component handle it.
What do you think? Any thought for this?

Thank you

**Comment by devnomic** [comment]
I'm wondering is it possible to change `useAtomValue` type declaration to make it only accept sync atom?
and create another `useAwaitedAtomValue` that only accept async atom, in the implementation it will call original `useAtomValue` that can accept promise atom?

**Comment by devnomic** [comment]
@dai-shi 
Could you please help to check the typing if this good enough?
https://codesandbox.io/s/lucid-fog-59s722?file=/src/useAwaitAtomValue.ts

Thank you

**Reply by dai-shi** [reply, marked as answer]
It looks good.
We don't change `useAtomValue`, so your solution on your end would be good.

**Reply by devnomic** [reply]
Thank you for your help, yes i'm quite happy for now since the editor now show error lint to prevent unintended behavior / bugs.

However this dx issue is something that i hope can be addressed officially in the future.

**Reply by dai-shi** [reply]
Well, in the future version of React, there will be no runtime error.
https://codesandbox.io/s/youthful-bose-3glmqd?file=/src/App.tsx
